### PR TITLE
DB-11046 DB-11107 Implement Splice DB SSO Authentication

### PR DESCRIPTION
**Pull request by ipraznik-splice**
# Descripton

The Single sign-on (SSO) authentication scheme allows a user to log in to the database authenticated through an external provider. Once authenticated by this provider, users can securely initiate one or more sessions in Splice database without having to log into database every time.

There are two new attributes in the driver:

authenticator - a type of token authenticator. Defines the authenticator to use for verifying user login credentials.

Possible values:

splice_jwt - symmetric security key is used from the splice keystore to sign and verify a token.

splice_jwt_pub - asymmetric key encryption scheme is used to sign a token. A public key from the splice keystore is used to verify a token.

splice_oauth - splice auth0 server is used to sign/verify a token

okta_oauth - okta oauth server is used to sign/verify a token

token - signed authentication JWT.

For example JDBC url:
jdbc:splice://localhost:1527/splicedb;authenticator=okta_oauth;token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIwMHUzM2hramZhRVdHQUtpYjVkNiIsIm5hbWUiOiJJZ29yIFByYXpuaWsiLCJlbWFpbCI6ImlwcmF6bmlrQHNwbGljZW1hY2hpbmUuY29tIiwiaXNzIjoiaHR0cHM6Ly9zcGxpY2VtYWNoaW5lLmNvbS9vYXV0aDIvZGVmYXVsdCIsImlhdCI6MTYwOTQxNDc0OSwiZXhwIjoxNjA5NDE4MzQ5LCJkYl91c2VyIjoiaWdvciJ9.UrcO4Z95HjHB4XAiijQWvQlcfAIRtTP8341Ewadg6f8

**The PR for EE part: https://github.com/splicemachine/spliceengine-ee/pull/597**

# How to test

https://splicemachine.atlassian.net/wiki/spaces/PD/pages/1765998642/How+to+Test+SSO+database+environment

### CODE REVIEW

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java [hunk 1]
```diff
@@ -175,6 +175,9 @@
     // Indicates Kerberos authentication.
     public static final int SECMEC_KERSEC = 0x0B;
 
+    // Indicates Token authentication.
+    public static final int SECMEC_TOKEN = 0x0E;
```
Comment (dgomezferro) [inline db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java:179]: Did you come up with this value or have you taken it from somewhere? As far as I could see the DRDA security mechanism that would work here is PLGIN (Plug-in security) which has value `0x0F`. `0x0E` stands for `EUSRNPWDDTA` (Encrypted User ID, Encrypted Password, Encrypted New Password, and Encrypted Security-sensitive Data)

Comment (dgomezferro) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java:179]: Check https://splicemachine.atlassian.net/wiki/spaces/PD/pages/581369866/DRDA+Specification+Documents for the DRDA specs

File: db-engine/pom.xml [hunk 2]
```diff
@@ -103,6 +103,29 @@
...
+            <artifactId>jjwt-jackson</artifactId> <!-- or jjwt-gson if Gson is preferred -->
+            <version>0.11.2</version>
+            <scope>runtime</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.fasterxml.jackson.core</groupId>
+                    <artifactId>jackson-databind</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
```
Comment (dgomezferro) [inline db-engine/pom.xml:128]: This should be moved to the -ee repo

File: db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     *                 splice.jwtToken.jksKeyStore
+     *                 splice.jwtToken.jksKeystore.password
+     *                 splice.jwtToken.jksKeystore.alias
+     */
+    protected KeyAccessTokenVerifier() {
+        keyStorePath = System.getProperty(SPLICE_JWT_TOKEN_JKS_KEY_STORE);
+        keyStorePassword = System.getProperty(SPLICE_JWT_TOKEN_JKS_KEYSTORE_PASSWORD);
+        keyStoreAlias = System.getProperty(SPLICE_JWT_TOKEN_JKS_KEYSTORE_ALIAS);
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java:33]: I don't think we want these as System properties, but rather as configuration parameters

File: db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    public String decodeUsername(String jwtToken) throws StandardException {
+        Key key = loadKey();
+        String userNameClaim = System.getProperty(AccessTokenVerifier.SPLICE_JWT_TOKEN_USERNAME);
+        assert StringUtils.isNotEmpty(userNameClaim);
+
+        Claims jwtBody = Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(jwtToken).getBody();
+
+        if (!jwtBody.get("iss").equals(issuer)) {
+            throw new IllegalArgumentException(String.format("Token is from another isser %s:%s", jwtBody.get("iss"), issuer));
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java:50]: Typo `isser` should be `issuer`

**Review by dgomezferro — CHANGES_REQUESTED** [review]

**Comment by dgomezferro** [comment]
Something else needed would be a `RemoteUserToken`, that class would be used to create connections to other RegionServers using the same authentication mechanism

**Review by jyuanca — CHANGES_REQUESTED** [review]

File: db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java (hunk 4 above)
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java]: fixed

File: db-engine/src/main/java/com/splicemachine/db/authentication/AccessTokenVerifier.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.splicemachine.db.iapi.error.StandardException;
+
+public interface AccessTokenVerifier {
+    String SPLICE_JWT_TOKEN_USERNAME = "splice.jwtToken.username";
+    String SPLICE_JWT_TOKEN_ISSUER = "splice.jwtToken.issuer";
+
+    String decodeUsername(String jwtToken) throws StandardException;
+
+}
```
[the remark this answers is not in the document]
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/authentication/AccessTokenVerifier.java:43]: added

File: db-engine/src/main/java/com/splicemachine/db/impl/token/AccessTokenVerifierFactory.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+                return new OktaAccessTokenVerifier();
+            case SPLICE_JWT:
+                return new KeyAccessTokenVerifier();
+            case SPLICE_JWT_PUB:
+                return new KeyPairAccessTokenVerifier();
+            default:
+                throw new IllegalArgumentException(String.format("%s authenticator is not supported", authenticatorName));
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/token/AccessTokenVerifierFactory.java:53]: added

File: db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+ * <p>
+ * This service instantiates and calls the basic User authentication scheme at runtime.
+ * </p>
+ *
+ * <p>
+ * User credentials are defined in the SYSUSERS table.
+ * </p>
+ */
+public final class TokenAuthenticationServiceImpl extends NativeAuthenticationServiceImpl {
+
```
[the remark this answers is not in the document]
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java:67]: It's allowed only one service in the configuration. To have the possibility to combine a token auth and native auth (username/password), it extends NativeAuthenticationServiceImpl and delegates to NativeAuthenticationServiceImpl.authenticateUser if username/password are passed.

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java (hunk 1 above)
Comment (ipraznik-splice) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java]: changed to use PLGIN. I've introduced the new constant 
`public static final int SECMEC_PLGIN = 0x0F;`
but still kept the old one with more meaning, that is used in the code:
`public static final int SECMEC_TOKEN = SECMEC_PLGIN;`

File: db-engine/pom.xml (hunk 2 above)
Comment (ipraznik-splice) [reply db-engine/pom.xml]: token implementation has been moved to splice-auth

File: db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java (hunk 3 above)
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/token/KeyAccessTokenVerifier.java]: changed to use a configuration. As I moved the code to EE/splice-auth, I needed to add some more dependencies to this module: hbase_storage, hadoop-common, hbase-client to use HConfiguration

**Comment by ipraznik-splice** [comment]
> Something else needed would be a `RemoteUserToken`, that class would be used to create connections to other RegionServers using the same authentication mechanism

I've created a new task for implementation: https://splicemachine.atlassian.net/browse/DB-11274

**Review by martinrupp — CHANGES_REQUESTED** [review]
- please add unit tests / ITs
- description of change should contain some small overview of the implementation and a short how to test this ( you can copy parts of https://splicemachine.atlassian.net/wiki/spaces/PD/pages/1530855455/Single+Sign-on+SSO+authentication i guess)

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2410/

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by ipraznik-splice** [comment]
> * please add unit tests / ITs
> * description of change should contain some small overview of the implementation and a short how to test this ( you can copy parts of https://splicemachine.atlassian.net/wiki/spaces/PD/pages/1530855455/Single+Sign-on+SSO+authentication i guess)

description is added.

the core functionality is in EE, there are JUnit and ITs: https://github.com/splicemachine/spliceengine-ee/pull/597. Functional auth ITs, which depend on the platform, are disabled now. Enabling of TOKEN authentication needs environment variables and some parameters in SpliceTestPlatformConfig. It's possible to run these ITs only against a preconfigured cluster currently, for example locally.

**Comment by martinrupp** [comment]
if i understand correctly, to test, i have to
- go to `platform_it/pom.xml`, profile `spliceFast`, replace `-Dsplice.authentication=NATIVE` by: `-Dsplice.authentication=TOKEN`
- in `spliceengine/platform_it/src/test/java/com/splicemachine/test/SpliceTestPlatformConfig.java`, below line `spliceengine/platform_it/src/test/java/com/splicemachine/test/SpliceTestPlatformConfig.java`, add the lines
```
config.set("splice.jwtToken.issuer", "https://splicemachine-dev.auth0.com/");
config.set("splice.jwtToken.audience", myJwtToken);
config.set("splice.jwtToken.auth0.cert", "/tmp/auth0.pub");
config.set("splice.jwtToken.username", "sub");
```

is that correct? What is that `/tmp/auth0.pub`, do i need to have that file?

can i somehow check that it then
- it works with jwt connection over sqlshell
- doesn't work if using jwt connection but wrong user
- same but wrong token
- works over other jdbc tools - maybe we can test ths later
- doesn't work with connecting with other auth methods (like BASIC)

as far as i know jwt tokens can expire, does the provided token expire (and if yes, would we need to recreate one, and how?).

**Review by martinrupp — CHANGES_REQUESTED** [review]
if i understand correctly, to test, i have to

- go to `platform_it/pom.xml`, profile `spliceFast`, replace `-Dsplice.authentication=NATIVE` by: `-Dsplice.authentication=TOKEN`
- in `spliceengine/platform_it/src/test/java/com/splicemachine/test/SpliceTestPlatformConfig.java`, below line 375 / `config.setBoolean("splice.replication.enabled", true);`, add the lines
```
config.set("splice.jwtToken.issuer", "https://splicemachine-dev.auth0.com/");
config.set("splice.jwtToken.audience", myJwtToken);
config.set("splice.jwtToken.auth0.cert", "/tmp/auth0.pub");
config.set("splice.jwtToken.username", "sub");
```

is that correct? What is that `/tmp/auth0.pub`, do i need to have that file?

Can you maybe create a branch with one commit of this changes so that testers can just cherry-pick that and be ready to test?

can i somehow check that it then
- it works with jwt connection over sqlshell
- doesn't work if using jwt connection but wrong user
- same but wrong token
- works over other jdbc tools - maybe we can test ths later
- doesn't work with connecting with other auth methods (like BASIC)

as far as i know jwt tokens can expire, does the provided token expire (and if yes, would we need to recreate one, and how?).

**Comment by martinrupp** [comment]
tested with dbeaver and in java using jdbc. sqlshell currently doesn't work yet.

**Comment by martinrupp** [comment]
ok had a zoom call, this is a bit more complicated to test, this will be done in the future.

**Review by jyuanca — APPROVED** [review]

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java [hunk 8]
```diff
@@ -2162,5 +2167,17 @@ protected void writeXATransactionStart(Statement statement) throws SqlException
     public void setServerPrincipal(String serverPrincipal) {
         this.serverPrincipal = serverPrincipal;
     }
+
+    @SuppressFBWarnings(value = "DM_DEFAULT_ENCODING", justification = "DB-11046")
+    private void flowUSRIDTokenconnect(String authenticator, String token) throws SqlException {
```
Comment (dgomezferro) [inline db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java:2172]: Isn't the "user id" embedded into the token? Why do we specify it separately too?

File: db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    public String authenticateUser (String userName, String userPassword, String databaseName, Properties info) throws SQLException {
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java:69]: I think we need a better mechanism to compose multiple authentication services at the same time. Can you open a Jira that removes the inheritance here and addresses the composability?

File: db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private AccessTokenVerifier getTokenVerifier(String authenticatorName) throws StandardException {
+        try {
+            Class<?> verifierFactoryClass = Class.forName("com.splicemachine.db.impl.token.AccessTokenVerifierFactory");
+            Method m = verifierFactoryClass.getDeclaredMethod("createVerifier", String.class);
+            return (AccessTokenVerifier) m.invoke(null, authenticatorName);
+        } catch (Exception e) {
+            throw StandardException.plainWrapException(e);
+        }
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java:133]: I don't like this reflection here. Would it be possible to use something similar to the BackupManager? A class in the -ee repo implements the AuthManager, a class in this repo throws UnsupportedOpEx.

Comment (dgomezferro) [reply db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java:133]: Or even move TokenAuthenticationServiceImpl to the -ee repo?

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java [hunk 11]
```diff
@@ -175,6 +175,13 @@
     // Indicates Kerberos authentication.
     public static final int SECMEC_KERSEC = 0x0B;
 
+    // The Plug-in Security Mechanism specifies that an external authentication mechanism is to be
+    // used to carry out the authentication.
+    public static final int SECMEC_PLGIN = 0x0F;
```
Comment (dgomezferro) [inline db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java:180]: Even though you are using the PLGIN code, I don't se we are following the DRDA protocol. Quoting from page 53:
> If the plug-in security mechanism (PLGIN) is returned as a SECMEC on the ACCSECRD reply data object, it must be accompanied by the list of supported plug-in modules in the PLGINLST instance variable following the ACCSECRD.

The token should be passed in through the PLGINSECTKN parameter (page 632)

An overview of the flows is described in PLGINOVR (page 627)

**Review by dgomezferro — CHANGES_REQUESTED** [review]

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java (hunk 8 above)
Comment (ipraznik-splice) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java:2172]: yes, it's embedded. It was done to simplify an implementation. If a user id not provided in parameters, it's filled as 'splice' by default and I can change it only after parsing of token, but as I remember there was a problem, that some parts were already initialized with 'splice' as user id. I have to look again to improve it.

File: db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java (hunk 10 above)
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java]: TokenAuthenticationServiceImpl has been moved to EE

File: db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java (hunk 9 above)
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java]: https://splicemachine.atlassian.net/browse/DB-11554

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java (hunk 8 above)
Comment (ipraznik-splice) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java:2192]: userId is taken from token now

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java (hunk 11 above)
Comment (ipraznik-splice) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConfiguration.java:180]: DRDA protocol has been implemented

File: db-drda/src/main/java/com/splicemachine/db/impl/drda/CodePoint.java [hunk 12]
```diff
@@ -875,8 +877,16 @@
...
 	static final int XIDCNT = 0x1906;
-	
-	
+
+
+	// Security Plug-in
+	static final int PLGINLSE = 0x1910;
+	static final int PLGINLST = 0x191E;
+	static final int PLGINNM = 0x191C;
+	static final int PLGINCNT = 0x190F;
```
Comment (martinrupp) [inline db-drda/src/main/java/com/splicemachine/db/impl/drda/CodePoint.java:886]: please describe a bit what these are doing so we don't have to scan the 772-page DRDA docu.
Either here or at another central place.
e.g.

```
// PLGINCNT: plugin count (int32 count)
static final int PLGINCNT = 0x190F;

// PLGINNM: plugin name (string name)
static final int PLGINNM = 0x191C;

/// PLGINLST: plugin list (PLGINCNT, count x (PLGINLSE) )
static final int PLGINLST = 0x191E;

// PLGINLSE: plugin list entry (PLGINNM, [PLGINPPL])
static final int PLGINLSE = 0x1910;
```

File: db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java [hunk 13]
```diff
@@ -3165,6 +3175,17 @@ else if (database.securityMechanism == CodePoint.SECMEC_USRSSBPWD)
...
         }
+
+        if (database.securityMechanism == CodePoint.SECMEC_PLGIN) {
+            writer.startDdm(CodePoint.PLGINLST);
+            writer.writeScalar2Bytes(CodePoint.PLGINCNT, CodePoint.SUPPORTED_PLUGINS.length);
+            for (int i=0; i < CodePoint.SUPPORTED_PLUGINS.length; i++) {
+                writer.startDdm(CodePoint.PLGINLSE);
+                writer.writeScalarString(CodePoint.PLGINNM, CodePoint.SUPPORTED_PLUGINS[i]);
+                writer.endDdm();
+            }
```
Comment (martinrupp) [inline db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java:3186]: extract to function `writePLGINLST()`

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java [hunk 14]
```diff
@@ -952,16 +956,27 @@ private void flowSecurityCheckAndAccessRdb(int securityMechanism,
...
+                                   byte[] encryptedPassword) throws SqlException {
+        flowSecurityCheck(securityMechanism, user, password, encryptedUserid, encryptedPassword, null);
+    }
+
+    private void flowSecurityCheck(int securityMechanism,
+                                   String user,
+                                   String password,
+                                   byte[] encryptedUserid,
+                                   byte[] encryptedPassword,
+                                   String pluginName
```
Comment (martinrupp) [inline db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java:971]: i would suggest to rename `pluginName` to `securityPlugin` or `securityPluginName` to make things clear in all code paths that this is about the security plugin.
please also add a documentation of this parameter. What are the allowed parameters for pluginName?

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java [hunk 15]
```diff
@@ -2105,6 +2113,30 @@ private void parseACCSECRD(NetConnection netConnection, int securityMechanism) t
...
+            List<String> pluginNames = new ArrayList<>(pluginListCount);
+            for (int i = 0; i < pluginListCount; i++) {
+                parseLengthAndMatchCodePoint(CodePoint.PLGINLSE);
+                parseLengthAndMatchCodePoint(CodePoint.PLGINNM);
+                pluginNames.add(readString());
+            }
+            peekCP = peekCodePoint();
+            popCollectionStack();
+            if (!pluginNames.contains(netConnection.authenticator.toUpperCase())) {
+                doSyntaxrmSemantics(CodePoint.SYNERRCD_REQ_OBJ_NOT_FOUND);
```
Comment (martinrupp) [inline db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java:2132]: can we have some log here? e.g. `"Trying to find " + netConnection.authenticator + ", but only have " + pluginNames.toString() `

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java [hunk 16]
```diff
@@ -2105,6 +2113,30 @@ private void parseACCSECRD(NetConnection netConnection, int securityMechanism) t
...
+            }
+        } else {
+            doSyntaxrmSemantics(CodePoint.SYNERRCD_REQ_OBJ_NOT_FOUND);
```
Comment (martinrupp) [inline db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java:2135]: please add a log

File: db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java [hunk 17]
```diff
@@ -1556,6 +1557,14 @@ private int getConnFromDatabaseName() throws DRDAProtocolException
             p.put(Attribute.DRDA_KERSEC_AUTHENTICATED, Boolean.toString(gssContext.isEstablished()));
         }
 
+        if (database.securityMechanism == CodePoint.SECMEC_TOKEN) {
+            p.put(Attribute.DRDA_SECMEC,
```
Comment (martinrupp) [inline db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java:1561]: can we log some things about this? like securityMechanism or pluginName?

**Review by martinrupp — CHANGES_REQUESTED** [review]
some small changes

File: db-drda/src/main/java/com/splicemachine/db/impl/drda/CodePoint.java (hunk 12 above)
Comment (ipraznik-splice) [reply db-drda/src/main/java/com/splicemachine/db/impl/drda/CodePoint.java]: added, to net/CodePoint.java as well

File: db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java (hunk 13 above)
Comment (ipraznik-splice) [reply db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java]: done

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java (hunk 14 above)
Comment (ipraznik-splice) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConnection.java:971]: done

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java (hunk 15 above)
Comment (ipraznik-splice) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java:2135]: There is no usage of free text messages for agent_.logWriter_ in the code. Added more trace via .traceEntry:
`agent_.logWriter_.traceEntry(this, "parsePLGINLST", netConnection.authenticator, pluginNames.toString());`

File: db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java (hunk 16 above)
Comment (ipraznik-splice) [reply db-client/src/main/java/com/splicemachine/db/client/net/NetConnectionReply.java:2141]: see above

File: db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java (hunk 17 above)
Comment (ipraznik-splice) [reply db-drda/src/main/java/com/splicemachine/db/impl/drda/DRDAConnThread.java:1561]: added logging for parsed authenticator and token. securityMechanism already logged, when it's parsed.

**Review by martinrupp — APPROVED** [review]

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2698/

**Pull request merged**